### ISSUE TITLE
Parallelize compilation

### ISSUE THREAD

**Issue by dadhi** [dadhi/FastExpressionCompiler#115]
Recently, I was thinking of a ways to speed up compilation using these 12x logical cores I have ;-)

But IL emitting is linear, except for __nested lambdas__.

They are compiled separately and in isolatation, actually starting a whole new `TryCompile` cycle.
 
What if we accumulate all the info for them and compile them in parallel.

It also has a separate benifit. Lambdas found and compiled during `TryCollectBoundConstants` step. Because they put in closure of container lambda. If the collect fails, we are wasting these nested compilation time.

So let's compile collected nested lambdas in parallel before emitting the expression code.

**Comment by jogibear9988** [comment]
do you really think it will speed up? it also has overhead involved, wich may be slower...

**Comment by dadhi** [comment]
I have hopes. I know about overhead, but test will tell. 

There is #72, where 5 or 6 nested lambdas. Will use it as a test.

**Comment by dadhi** [comment]
@dzmitry-lahoda 

The compilation itself (aka `DynamicMethod.CreateDelegate`) is the heaviest thing of all. I have a benchmark somewhere in benchmarks where I measure FEC.FastCompile against completely by-hand Emit and CreateDelegate (where I know what to Emit - no loops, not collect stages, no FEC infrastructure overheads). The result difference is only 5 to 10% (FEC stuff), the rest 90% is `CreateDelegate`. That's why the investing time to optimize FEC code won't provide "significant gains. But it does not mean I won't do it too :-) 

On the other hand, compiling each nested lambda does the `CreateDelegate`. So cutting off `CreateDelegate` total time by doing them in parallel seems more promising. The test will tell.

Regarding implementation, I want to try `Task.Run`, `Task.WaitAll` first, cause they are supported for all FEC target platforms.

**Comment by dadhi** [comment]
Huh, the benchmark was in DryIoc. Have added it to FEC.

Check below that FEC.FastCompile is __3.25__ vs __3.11__ for Emit/CreateDelegate by hand

![image](https://user-images.githubusercontent.com/39516/44842878-a70f6d00-ac4f-11e8-85af-c36cb6b2e305.png)

**Comment by dzmitry-lahoda** [comment]
TaskRun_Wait_ManualEmit?

**Comment by dadhi** [comment]
@dzmitry-lahoda 
> TaskRun_Wait_ManualEmit?
Sorry, didn't get it.

Here is the benchmark: https://github.com/dadhi/FastExpressionCompiler/blob/master/test/FastExpressionCompiler.Benchmarks/FEC_vs_ManualEmit_vs_Activator.cs

**Comment by dzmitry-lahoda** [comment]
Points for grows of FEC stuff:
1. Full support of LINQ expression supported
2. Extended support of ExpressionInfo only features
3. Heuristics to generate better more optimal IL for task at hand.
4. Very asked request to do some validation (may be throws/return different errors depending on reason why FEC failed,  on expression level and IL level https://github.com/kevin-montrose/Sigil validation) . But it can be made optional and used via parameter only in debug build.

May be long time before FEC reach much above delegate compilation. Question will be how often lambda has child lambda.

**Comment by dadhi** [comment]
> 4. Very asked request to do some validation (may be throws/return different errors depending on reason why FEC failed, on expression level and IL level https://github.com/kevin-montrose/Sigil validation) . But it can be made optional and used via parameter only in debug build.

This issue: #53

**Comment by dzmitry-lahoda** [comment]
```
// tries to measure performance overhead of thread in worth case when no need for thread
 [Benchmark]
public object TaskRun_Wait_ManualEmit()
{
    Task.Run(ManualEmit).Wait();
}
```

**Comment by dadhi** [comment]
> Task.Run(ManualEmit).Wait();

Make sense to check.

**Comment by dadhi** [comment]
@dzmitry-lahoda 
Here the results:

![image](https://user-images.githubusercontent.com/39516/44848517-b7c7df00-ac5f-11e8-8ea5-ff0c1acbc5c9.png)

The overhead is 2x. But it surely depends on expression (here it is quite simple without closure).

`ObjectMethodExecutor` should be much closer to a real case, which I want to address.

Likely, in order to prevent regressions, I will introduce separate overload: 

`async Task<TDelegate> CompileFastWithCompilingNestedLambdasInParallel` or something,

**Btw,** we may explore other ways of parallel execution beside the `Task`.

**Comment by jogibear9988** [comment]
Maybe there is a way to improve speed of DynamicMethod.CreateDelegate, or does it raise unmanned Code?

**Comment by dadhi** [comment]
Maybe, always go deeper :)

Someday I will decompile and look.

**Comment by dadhi** [comment]
Here is the result from _ParallelNestedLambda_ branch:

![image](https://user-images.githubusercontent.com/39516/45082774-c2242600-b102-11e8-808d-dd0e4b034621.png)

**Comment by dadhi** [comment]
Closing, no go for now.
